core/src/main/java/org/sonarsource/sonarlint/core/tracking/SecurityHotspotMatchingServiceImpl.java

```diff
@@ -0,0 +1,165 @@
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.sonarsource.sonarlint.core.clientapi.backend.hotspot.HotspotStatus;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.ClientTrackedFindingDto;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.LocalOnlySecurityHotspotDto;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.MatchWithServerSecurityHotspotsParams;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.MatchWithServerSecurityHotspotsResponse;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.SecurityHotspotMatchingService;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.ServerMatchedSecurityHotspotDto;
+import org.sonarsource.sonarlint.core.commons.Binding;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+import org.sonarsource.sonarlint.core.issuetracking.Trackable;
+import org.sonarsource.sonarlint.core.issuetracking.Tracker;
+import org.sonarsource.sonarlint.core.repository.config.ConfigurationRepository;
+import org.sonarsource.sonarlint.core.repository.vcs.ActiveSonarProjectBranchRepository;
+import org.sonarsource.sonarlint.core.serverapi.hotspot.ServerHotspot;
+import org.sonarsource.sonarlint.core.serverconnection.StorageService;
+import org.sonarsource.sonarlint.core.sync.SynchronizationServiceImpl;
+
+import static org.sonarsource.sonarlint.core.utils.FutureUtils.waitForTask;
+import static org.sonarsource.sonarlint.core.utils.FutureUtils.waitForTasks;
+
+@Named
+@Singleton
+public class SecurityHotspotMatchingServiceImpl implements SecurityHotspotMatchingService {
+  private static final int FETCH_ALL_SECURITY_HOTSPOTS_THRESHOLD = 10;
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+  private final ConfigurationRepository configurationRepository;
+  private final StorageService storageService;
+  private final ActiveSonarProjectBranchRepository activeSonarProjectBranchRepository;
+  private final SynchronizationServiceImpl synchronizationService;
+  private final ExecutorService executorService;
+
+  public SecurityHotspotMatchingServiceImpl(ConfigurationRepository configurationRepository, StorageService storageService,
+    ActiveSonarProjectBranchRepository activeSonarProjectBranchRepository, SynchronizationServiceImpl synchronizationService) {
+    this.configurationRepository = configurationRepository;
+    this.storageService = storageService;
+    this.activeSonarProjectBranchRepository = activeSonarProjectBranchRepository;
+    this.synchronizationService = synchronizationService;
+    this.executorService = Executors.newSingleThreadExecutor(r -> new Thread(r, "sonarlint-server-tracking-hotspot-updater"));
+  }
+
+  @Override
+  public CompletableFuture<MatchWithServerSecurityHotspotsResponse> matchWithServerSecurityHotspots(MatchWithServerSecurityHotspotsParams params) {
+    return CompletableFutures.computeAsync(cancelChecker -> {
```

Comment:
Do we really need to depends on lsp4j cancellation mechanism? Do we support cancellation here?

Reply:
I think it makes sense for this step to be cancelable as the network is involved. lsp4j is not a good way to achieve that but I suggest we keep it and reconsider it when we move to a separate process